app/bundles/EmailBundle/Tests/Controller/EmailMapStatsControllerTest.php

```diff
@@ -0,0 +1,287 @@
+        $this->em->persist($role);
+        $this->em->flush();
+
+        $user = new User();
+        $user->setFirstName('Example');
+        $user->setLastName('Example');
+        $user->setUsername('Example');
+        $user->setPassword('123456');
+        $user->setEmail('example@example.com');
+        $user->setRole($role);
+        $this->em->persist($user);
+        $this->em->flush();
+
+        $email = new Email();
+        $email->setName('Test email 1');
+        $email->setCreatedBy($user);
+        $this->em->persist($email);
+        $this->em->flush();
+
+        $corePermissionsMock->method('hasEntityAccess')
+            ->with(
+                'email:emails:viewown',
+                'email:emails:viewother',
+                $user->getId()
+            )
+            ->willReturn(false);
+
+        $result = $this->mapController->hasAccess($corePermissionsMock, $email);
+
+        try {
+            $this->mapController->viewAction($corePermissionsMock, $email->getId(), '2023-07-20', '2023-07-27');
+        } catch (AccessDeniedHttpException|\Exception $e) {
+            $this->assertTrue($e instanceof AccessDeniedHttpException);
+        }
+
+        $this->assertFalse($result);
+    }
+
+    /**
+     * @throws OptimisticLockException
+     * @throws ORMException
+     */
+    public function testGetMapOptions(): void
+    {
+        $email = new Email();
+        $email->setName('Test email');
+        $this->em->persist($email);
+        $this->em->flush();
+
+        $result = $this->mapController->getMapOptions();
+        $this->assertSame(EmailMapStatsController::MAP_OPTIONS, $result);
+    }
+
+    /**
+     * @return array<string, array<int, array<string, string>>>
+     */
+    private function getStats(): array
+    {
+        return [
+            'clicked_through_count' => [
+                [
+                    'clicked_through_count' => '4',
+                    'country'               => '',
+                ],
+                [
+                    'clicked_through_count' => '7',
+                    'country'               => 'Italy',
+                ],
+            ],
+            'read_count' => [
+                [
+                    'read_count'            => '4',
+                    'country'               => '',
+                ],
+                [
+                    'read_count'            => '12',
+                    'country'               => 'Italy',
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function testGetData(): void
+    {
+        $email = new Email();
+        $email->setName('Test email');
+
+        $dateFrom = new \DateTimeImmutable('2023-07-20');
+        $dateTo   = new \DateTimeImmutable('2023-07-25');
+
+        $this->emailModelMock->method('getCountryStats')
+            ->with($email, $dateFrom, $dateTo, false)
+            ->willReturn($this->getStats());
```

Comment:
You don't need mocks. Instead of this store the Stats to the database and let it select them live. This way you'll test that the query works as you expect.

Reply:
Ok, I removed the mocks.